### ISSUE TITLE
Cannot set order of filters with citeproc in YAML defaults

### ISSUE THREAD

**Issue by sdbbs** [jgm/pandoc#7230]
Consider this `mwe.md`:

```
---
title: My minimal working example
author: John Doe
numbersections: true    # else pandoc-crossref references are missing
link-citations: true
linkcolor: red
bibliography: mwe.bib
...

# Introduction { #sec:intro }

Hello world

# Contents

As it is mentioned in section [@sec:intro], Pandoc is great [@bib:pandoc].

```

... with this `mwe.bib`:

```
@misc{bib:pandoc,
  title = {Pandoc: A Universal Document Converter},
  howpublished = {http://johnmacfarlane.net/pandoc},
}

```

Note that [pandoc-crossref docs](https://github.com/lierdakil/pandoc-crossref/blob/master/docs/index.md) mention:

> Since pandoc-crossref uses the same citation syntax as citeproc, you have to run former before latter. For example:
>
>     pandoc -F pandoc-crossref --citeproc file.md -o file.html

This works in this example also -- from the command line; for `pandoc-crossref` before `citeproc`, all seems fine:

```
$ pandoc -F pandoc-crossref --citeproc --pdf-engine=xelatex mwe.md -o mwe.pdf

$ pandoc --verbose -F pandoc-crossref --citeproc --pdf-engine=xelatex mwe.md -o mwe.pdf 2>&1 | head -n 4
[INFO] Running filter pandoc-crossref
[INFO] Completed filter pandoc-crossref in 0 ms
[makePDF] temp dir:
D:/msys64/tmp/tex2pdf.-135552d685436b73
``` 

... nowever, if I invert the order, I get an expected warning from `citeproc`:

```
$ pandoc --citeproc -F pandoc-crossref --pdf-engine=xelatex mwe.md -o mwe.pdf
[WARNING] Citeproc: citation sec:intro not found

$ pandoc --verbose --citeproc -F pandoc-crossref --pdf-engine=xelatex mwe.md -o mwe.pdf 2>&1 | head -n 4
[WARNING] Citeproc: citation sec:intro not found
[INFO] Running filter pandoc-crossref
[INFO] Completed filter pandoc-crossref in 15 ms
[makePDF] temp dir:
```

The verbose output in both cases gives a hint that `citeproc` is not being considered a filter. But in any case, setting order from the command line works. However, there are still problems in the PDF output:

![image](https://user-images.githubusercontent.com/48317456/115150998-3b31db00-a06b-11eb-9a94-9462aa1c23da.png)

In other words, -- even if the warning is about an intern cross-reference, -- the bibliography entry is rendered in bibliography - but the citation to it is *not* rendered (while the cross-reference the warning is about *is* correctly rendered).

Now let's try with this `mwe.yaml`:

```yaml
pdf-engine: xelatex
bibliography: mwe.bib
filters:
- pandoc-crossref
- citeproc
```

We get no warnings:

```
$ pandoc --verbose --defaults=mwe.yaml mwe.md -o mwe.pdf 2>&1 | head -n 4
[INFO] Running filter pandoc-crossref
[INFO] Completed filter pandoc-crossref in 15 ms
[makePDF] temp dir:
D:/msys64/tmp/tex2pdf.-988567037aa91979
```

,,,, however, `citeproc` has not run, and the bib entry in document is not parsed.

Let's now add `citeproc: true` to `mwe.yaml`:

```yaml
pdf-engine: xelatex
citeproc: true
bibliography: mwe.bib
filters:
- pandoc-crossref
- citeproc
```

We get the warning this time:

```
$ pandoc --verbose --defaults=mwe.yaml mwe.md -o mwe.pdf 2>&1 | head -n 4
[WARNING] Citeproc: citation sec:intro not found
[INFO] Running filter pandoc-crossref
[INFO] Completed filter pandoc-crossref in 15 ms
[makePDF] temp dir:
```

... which implies `citeproc` has run before `pandoc-crossref` - even if `citeproc`  sits *after*  `pandoc-crossref` in the `filters:` list.

Let's try move `citeproc: true` as last in the `mwe.yaml`:

```yaml
pdf-engine: xelatex
bibliography: mwe.bib
filters:
- pandoc-crossref
- citeproc
citeproc: true
```

Now we get:

```
$ pandoc --verbose --defaults=mwe.yaml mwe.md -o mwe.pdf 2>&1 | head -n 4
[WARNING] Citeproc: citation sec:intro not found
[INFO] Running filter pandoc-crossref
[INFO] Completed filter pandoc-crossref in 0 ms
[makePDF] temp dir:
```

... the same warning as before, which implies `citeproc` has again run first/before `pandoc-crossref`.

Using `cite-method: citeproc` instead of `citeproc: true` has the same effect, as not having `citeproc` enabled at all.

Using `- pandoc-citeproc` in `filters:` instead of `- citeproc`, so `mwe.yaml` like this:

```yaml
pdf-engine: xelatex
bibliography: mwe.bib
cite-method: citeproc
filters:
- pandoc-crossref
- pandoc-citeproc
```

... does seem to set the order, but generates an error otherwise:

```
$ pandoc --verbose --defaults=mwe.yaml mwe.md -o mwe.pdf 2>&1 | head -n 7
[WARNING] Deprecated: pandoc-citeproc filter. Use --citeproc instead.
[INFO] Running filter pandoc-crossref
[INFO] Completed filter pandoc-crossref in 0 ms
[INFO] Running filter pandoc-citeproc
pandoc-citeproc: Error in $: Incompatible API versions: encoded with [1,22] but attempted to decode with [1,21].
CallStack (from HasCallStack):
  error, called at .\Text\Pandoc\JSON.hs:112:48 in pandoc-types-1.21-7Y4X1UYmiAA6TrmyVvLJqQ:Text.Pandoc.JSON
```

And, using `- citeproc: true` instead of `- citeproc` under `filters:` in YAML causes "Error parsing mwe.yaml"

While ordering works from the command line - Is there any way to set the order of filters, in which `pandoc-crossref` runs before `citeproc` - strictly from the YAML file?

**Comment by jgm** [comment]
It is supposed to be possible to control this in the filters list...so this looks like a bug.
You're using the latest release, I assume?

**Comment by sdbbs** [comment]
Thanks @jgm,

>  You're using the latest release, I assume?

Uuups, sorry, forgot that:

```
$ pandoc --version
pandoc.exe 2.11.4
Compiled with pandoc-types 1.22, texmath 0.12.1, skylighting 0.10.2,
citeproc 0.3.0.5, ipynb 0.1.0.1
```

-----------

Btw, just wanted to post a workaround, as it took me a while to figure out:

So, if you have to use only `-F pandoc-crossref` and `--citeproc` on the command line so as to set order, but would otherwise like to keep everything else in .yaml, here is a variant that works:

`mwe.md`:

```
---
title: My minimal working example
author: John Doe
#link-citations: true # works here -- but not in yaml, unless it is in `metadata:` section!

...

# Introduction { #sec:intro }

Hello world

# Contents

As it is mentioned in section [@sec:intro], Pandoc is great [@bib:pandoc].


```

`mwe.bib` same as above;

`mwe.yaml`:

```yaml
pdf-engine: xelatex
bibliography: mwe.bib
#link-citations: true # Unknown option "link-citations"
## skip the below ones, if setting order from command line (else e.g. pandoc-crossref will run twice)
#cite-method: citeproc
#citeproc: true
#filters:
#- pandoc-crossref
#- citeproc
metadata:
  link-citations: true    # works fine here!
variables:
  #link-citations: true    # passes here, but no effect on citeproc
  numbersections: true    # else pandoc-crossref references are missing
  linkcolor: red

```

Command line in that case is:

```
pandoc --defaults=mwe.yaml -F pandoc-crossref --citeproc mwe.md -o mwe.pdf
```

**Comment by jgm** [comment]
Can you try with the latest release? I wasn't able to reproduce this.
Using
```
pdf-engine: xelatex
bibliography: mwe.bib
filters:
- pandoc-crossref
- citeproc
```
and `pandoc -d mwe.yaml mwe.md -o mwe.pdf` everything worked fine, and the bibliography was produced.

**Comment by sdbbs** [comment]
Many thanks, @jgm,

I have just updated [^1] `pandoc` - and updated to corresponding `pandoc-crossref`:

```
$ pandoc --version
pandoc.exe 2.13
Compiled with pandoc-types 1.22, texmath 0.12.2, skylighting 0.10.5,
citeproc 0.3.0.9, ipynb 0.1.0.1

$ pandoc-crossref --version
pandoc-crossref v0.3.10.0 git commit 2b62c009b52ea8b65e58666fe2fa0c8963a5002e (HEAD) built with Pandoc v2.13, pandoc-types v1.22 and GHC 8.10.4
```

,,,, and used this `mwe.yaml`:

```yaml
pdf-engine: xelatex
bibliography: mwe.bib
metadata:
  link-citations: true    # works fine here!
variables:
  numbersections: true    # else pandoc-crossref references are missing
  linkcolor: red
filters:
- pandoc-crossref
- citeproc
```

And indeed, as you mentioned, with:

> `pandoc -d mwe.yaml mwe.md -o mwe.pdf` everything worked fine, and the bibliography was produced

... which is great! Thanks, I will be closing this ....

I just wanted to mention something: I checked the `pandoc` run with `--verbose`:

```
$ pandoc --verbose --defaults=mwe.yaml mwe.md -o mwe.pdf
[INFO] Running filter pandoc-crossref
[INFO] Completed filter pandoc-crossref in 0 ms
[makePDF] temp dir:
....
```

I think the "[INFO] Running filter" report part is great, as it allows a sanity check for the order of running of filters; however, I can notice, `citeproc` is not included in that report. It could be that `citeproc` is no longer considered a "filter" - but could you reconsider adding a log line for `citeproc` in the "[INFO] Running filter" section as well? It would assist troubleshooting greatly ...

--------

[^1]: While I was trying to update, I noticed that I had `pandoc-citeproc.exe` in my `C:\Program Files\Pandoc`; so my previous state (before update) was:

```
$ pandoc --version
pandoc.exe 2.11.4
Compiled with pandoc-types 1.22, texmath 0.12.1, skylighting 0.10.2,
citeproc 0.3.0.5, ipynb 0.1.0.1
...

$ pandoc-crossref --version
pandoc-crossref v0.3.9.1 git commit ffd548d77c801075b4274de29a58bce1faef6872 (HEAD) built with Pandoc v2.11.4, pandoc-types v1.22 and GHC 8.10.3

$ pandoc-citeproc --version
pandoc-citeproc 0.17.0.1
```

... which I found kinda strange, because I remember at a certain point, citeproc was included into pandoc itself.

Then I checked Windows installed programs list, and was surprised to see this:

![2021-04-19_Programs and Features_pandoc](https://user-images.githubusercontent.com/48317456/115176088-a8c22380-a0cc-11eb-88d0-e225d22f6ce4.png)

Basically, I had always just ran the installer to updated (without previousl uninstall) - and seemingly, Windows kept tabs on all of those installs. My guess is, since the entire installation is just a few files that include `pandoc.exe`, just copying them in the installation dir so they overwrite earlier versions works fine - but that also resulted with a "hanging" `pandoc-citeproc.exe` from the time it was separate.

So it could be, that the ordering problem I was seeing, may have been due to an earlier `pandoc-citeproc.exe` sitting in the install folder?! Still, great that it is an irrelevant problem now, since after this update, all works for me, and I'm happy :)

Thanks again for the help!

**Comment by jgm** [comment]
on this see #6976
